docs/proposals/015-bindless-t.md

```diff
@@ -0,0 +1,199 @@
+to provide their own implementation of `getResourceFromBindlessHandle` to prevent breakages.
+
+If the user application requires a different bindless implementation, this default behavior can be overrided by defining
+`getResourceFromBindlessHandle` in the user code. Below is a possible user-space implementation of `getResourceFromBindlessHandle`
+for Vulkan:
+
+```slang
+
+// All texture and buffer handles are defined in descriptor set 100.
+[vk::binding(0, 100)]
+__DynamicResource<__DynamicResourceKind.General> resourceHandles[];
+
+// All sampler handles are defined in descriptor set 101.
+[vk::binding(0, 101)]
+__DynamicResource<__DynamicResourceKind.Sampler> samplerHandles[];
+
+export getResourceFromBindlessHandle<T>(uint64_t handle) where T : IOpaqueHandle
+{
+    if (T.kind == ResourceKind.Sampler)
+        return (T)samplerHandles[handle];
+    else
+        return (T)resourceHandles[handle];
+}
+```
+
+### Invalid Handle
+
+We reserve `uint64_t.maxValue` as a special handle value of `Bindless<T>` types to mean an invalid/null resource.
+This will allow us to optimize `Optional<Bindless<Texture2D>>` to use the reserved value to mean no-value.
+
+The user should also be able to use `Bindless<T>.invalid` to refer to such an invalid value:
+
+```slang
+struct Bindless<T> where T:IOpaqueHandle
+{
+    static const Bindless<T> invalid = Bindless<T>(uint64_t.maxValue);
+}
+```
+
+## Alternatives Considered
+
+We initially considered to support a more general `Bindless<T>` where `T` can be any composite type, for example, allowing the following:
+
+```slang
+struct Foo
+{
+    Texture2D t;
+    SamplerState s;
+    float ordinaryData;
+}
+
+uniform Bindless<Foo> foo;
+```
+
+which is equivalent to:
+
+```slang
+struct Bindless_Foo
+{
+    Bindless<Texture2D> t;
+    Bindless<SamplerState> s;
+    float s;
+}
+uniform Bindless_Foo foo;
+```
+
+While relaxing `T` this way adds an extra layer of convenience, it introduces complicated
+semantic rules to the type system, and there is increased chance of exposing tricky corner
+cases that are hard to get right.
+
+An argument for allowing `T` to be general composite types is that it enables sharing the same
+code for both bindless systems and bindful systems. But this argument can also be countered by
+allowing the compiler to treat `Bindless<T>` as `T` in a special mode if this feature is found to be useful.
+
+For now we think that restricting `T` to be an `IOpaqueHandle` type will result in a much simpler implementation, and is likely sufficient for current needs. Given that the trend of modern GPU architecture is moving towards bindless idioms and the whole idea of opaque handles may disappear in the future, we should be cautious at inventing too many heavy weight mechanisms around opaque handles. Nevertheless, this proposal still allows us to relax this requirement in the future if it becomes clear that such feature is valuable to our users.
```

Comment:
I’m not sure where to put this, but figured I’d ask, is there a way to specify that a bindless structured buffer is “globally coherent”? 

This appears to be a limitation with slang’s buffer pointers, so I’m currently carrying buffers around in a single mutable descriptor array, where I then create an unofficial integer addressing like described here. Is the idea to have the user add this via what’s retrieved from getResourceFromBindlessHandle?

Reply:
This will need to be a separate change that is not related to this proposal. We should probably extend StructuredBuffer and ByteAddressBuffer to take one additional generic parameter to specify the memory semantic, just like how we extended the types to include layout.